### DISCUSSION TITLE
Supported property for function in latest jsbsim plugin for Unreal

### DISCUSSION THREAD

**Discussion by nmfarras** [JSBSim-Team/jsbsim#1046] [Q&A]
I noticed that some `<property>` are not supported in Unreal Engine and may cause crash. For example in following function

```xml
    <function name="aero/coefficient/Cz_g">
      <description>
            Change in Z due to ground effect

            This is bad hack, GE should really be a addition, not a factor
        </description>
        <sum>
            <v>1</v>
            <product>
                <!-- positive contribution to lift -->
                <property>aero/coefficient/CL_g</property>
                <cos>
                  <!-- Less lift due to ground effect when flying inverted-->
                  <product>
                      <p>/orientation/roll-deg</p>
                      <v>0.017453</v>
                  </product>
                </cos>
                <cos>
                    <property>aero/alpha-rad</property>
                </cos>
            </product>
            <product>
                <!-- negative contribution to induced drag -->
                <v>-1</v>
                <property>aero/coefficient/CL_g</property>
                <cos>
                  <!-- Less lift due to ground effect when flying inverted-->
                  <product>
                      <p>/orientation/roll-deg</p>
                      <v>0.017453</v>
                  </product>
                </cos>
                <sin>
                    <property>aero/alpha-rad</property>
                </sin>
            </product>
        </sum>
    </function>
```
`/orientation/roll-deg` may cause the Unreal to crash. How can i check which property supported and which one not? I try this from f16 flight gear aerodynamic file.

**Comment by gallonmate** [comment]
Good question. I don't think there's any easy way to know which property will work or not, besides just testing it. FlightGear has a scripting system called Nasal. That property `p>/orientation/roll-deg</p>`is created from the F16's Nasal file. JSBSim does not use Nasal, and then JSBSim/Unreal plugin crashes when it can't find the property. 

It's possible to create the property name in the aircraft file so that it does exist and won't crash, (but I'm unfamiliar with how to correctly do that and can't help you). The other option is you can remove those properties/systems.

**Reply by nmfarras** [reply]
If by any chance you know from which file the definition of property is declare, I would love to know. Cause https://github.com/JSBSim-Team/jsbsim/discussions/1045#discussion-6255504 might have the same problem as me.

**Reply by nmfarras** [reply]
https://github.com/JSBSim-Team/jsbsim/tree/master/data_output I just check this folder by chance just now. Does this is the only property available to get and set with? Or there is more property available?

**Reply by nmfarras** [reply]
Just checking to be sure, does this can be used as reference?

https://github.com/JSBSim-Team/jsbsim/blob/3d95aeb751e6ddd11b1007571b86662cd41612d8/src/models/FGPropagate.cpp#L880-L882

For example if i want to make function using roll orientation parameter in degree, i can use `<p>attitude/phi-rdeg</p>` instead for the code I used in the question as subtitute?

**Reply by gallonmate** [reply]
> https://github.com/JSBSim-Team/jsbsim/tree/master/data_output I just check this folder by chance just now. Does this is the only property available to get and set with? Or there is more property available?

No, data_output doesn't appear to be a full list. There are around over 600 properties when a regular aircraft is loaded, you can see an example of this when you use the jsbsim component blueprint node "Property Manager Get Catalog" to see the list of properties after loading an aircraft. But even if you had a full list of the regular properties available in JSBSim, the problem is that JSBSim can create new properties at runtime with the aircraft files. So using a known list could create false positives on aircraft that might work fine. 

For example the airplane c172x has an autopilot file c172ap.xml and it creates new properties at runtime which don't exist in JSBSim code. JSBSim loads the c172x file, which then loads the autopilot file and new properties are created. Those properties are used in the c172x file and cannot be found the JSBSim code before the autopilot file is loaded. But it works fine in UE5. Similar thing with the AH1S helicopter. It loads a custom rotor system with new properties and it also works fine in UE5.

**Reply by gallonmate** [reply]
> Just checking to be sure, does this can be used as reference?
> 
> https://github.com/JSBSim-Team/jsbsim/blob/3d95aeb751e6ddd11b1007571b86662cd41612d8/src/models/FGPropagate.cpp#L880-L882
> 
> For example if i want to make function using roll orientation parameter in degree, i can use `<p>attitude/phi-rdeg</p>` instead for the code I used in the question as subtitute?

Yes those type of properties can be referenced. But I cannot say yes or no if it will work as substitute for roll orientation.

**Reply by nmfarras** [reply]
> For example the airplane c172x has an autopilot file c172ap.xml and it creates new properties at runtime which don't exist in JSBSim code. JSBSim loads the c172x file, which then loads the autopilot file and new properties are created. Those properties are used in the c172x file and cannot be found the JSBSim code before the autopilot file is loaded. But it works fine in UE5. Similar thing with the AH1S helicopter. It loads a custom rotor system with new properties and it also works fine in UE5.

Ah I see. I kinda get that you can create custom property but I just wanna know the default one for now. @jonsberndt answer shows what you mention since the catalog summarize the property used by the model, and i notice it is different between aircraft. Do I just need to exclude the custom one to get all the default jsbsim property?

**Comment by jonsberndt** [comment]
Not sure if it might be helpful in this particular circumstance, but executing this statement (for example):

./JSBSim --catalog=c172x

will print a list of all published properties for the model.

**Reply by nmfarras** [reply]
That is quite helpful! But I guess it include the user defined property and jsbsim default property. Do you know what the (R) and (RW) indicate?

```
    ic/vw-north-fps (R)
    ic/vw-east-fps (R)
    ic/vw-down-fps (R)
    ic/vw-mag-fps (R)
    ic/vw-dir-deg (RW)
    ic/roc-fps (RW)
    ic/u-fps (RW)
    ic/v-fps (RW)
    ic/w-fps (RW)
```

**Comment by seanmcleod** [comment]
R - Readable
W - Writeable

So writeable properties can be written to/updated by JSBSim script code, Python code, Nasal script, C++ code etc.

**Comment by seanmcleod** [comment]
> I just wanna know the default one for now

Well, a comprehensive default list is a bit tricky. So yes, the properties you see in the JSBSim C++ source code being added to the property tree are available by default, mostly.

https://github.com/JSBSim-Team/jsbsim/blob/3d95aeb751e6ddd11b1007571b86662cd41612d8/src/models/FGPropagate.cpp#L880-L882

I say mostly because you will also find properties in the C++ code like the following, that are only available if a piston engine for example is being used.

https://github.com/JSBSim-Team/jsbsim/blob/3d95aeb751e6ddd11b1007571b86662cd41612d8/src/models/propulsion/FGPiston.cpp#L375-L380

**Reply by nmfarras** [reply]
> > https://github.com/JSBSim-Team/jsbsim/blob/3d95aeb751e6ddd11b1007571b86662cd41612d8/src/models/FGPropagate.cpp#L880-L882
> > 
> > For example if i want to make function using roll orientation parameter in degree, i can use `<p>attitude/phi-rdeg</p>` instead for the code I used in the question as subtitute?
> 
> Yes those type of properties can be referenced. But I cannot say yes or no if it will work as substitute for roll orientation.

While at it i want to know does `<p>attitude/phi-deg</p>` can be used to refer roll orientation?
